ultravox/data/datasets.py

```diff
@@ -550,77 +549,48 @@ def __len__(self):
         return self._length
 
 
-class StopStrategy(str, enum.Enum):
-    FIRST_EXHAUSTED = "FIRST_EXHAUSTED"
-    LAST_EXHAUSTED = "LAST_EXHAUSTED"
-    NEVER_STOP = "NEVER_STOP"
-
-
 class InterleaveDataset(SizedIterableDataset):
-    """Interleaves multiple IterableDataset objects based on normalized weights."""
+    """Interleaves multiple SizedIterableDataset objects based on normalized weights."""
 
     def __init__(
         self,
         datasets: Sequence[SizedIterableDataset],
         weights: Optional[Sequence[float]] = None,
-        stop_strategy: StopStrategy = StopStrategy.LAST_EXHAUSTED,
-        seed: Optional[int] = 42,
-        static: bool = False,
     ) -> None:
         """
         Args:
             datasets: A list of SizedIterableDataset objects.
-            weights: A list of weights for each dataset.
-            stop_strategy: Strategy for stopping iteration.
+            weights: An optional list of dataset weights, i.e., the number of times it should be repeated.
             seed: Optional seed for reproducibility.
-            static: If true, the datasets are interleaved in a static order with equal weights.
         """
         self._datasets = datasets
-        self._rng = np.random.default_rng(seed)
-        self._static = static
-        self._stop_strategy = stop_strategy
-
-        if weights is None:
+        if weights is not None:
+            assert len(weights) == len(datasets)
+        else:
             weights = [1.0] * len(datasets)
-        total_weight = sum(weights)
-        self._normalized_probs = [w / total_weight for w in weights]
+        self._weighted_samples = [int(w * len(d)) for w, d in zip(weights, datasets)]
+        self._total_samples = sum(self._weighted_samples)
 
     def __iter__(self):
-        # If no datasets are provided, return an empty iterator
-        if not self._datasets:
-            return
-
-        iters = [iter(ds) for ds in self._datasets]
-        exhausted = [False] * len(iters)
-
-        if self._static:
-            static_iter = itertools.cycle(range(len(self._datasets)))
-
-        while True:
-            if self._static:
-                iter_index = next(static_iter)
-            else:
-                iter_index = self._rng.choice(len(iters), p=self._normalized_probs)
-
+        ds_iters = [iter(ds) for ds in self._datasets]
+        ds_pos = [0] * len(ds_iters)
+        # Find the iterator that is least far along and vend from it.
+        for i in range(self._total_samples):
+            min_fraction = 1.0
+            for j in range(len(ds_iters)):
+                iter_fraction = ds_pos[j] / self._weighted_samples[j]
+                if iter_fraction < min_fraction:
+                    min_fraction = iter_fraction
+                    iter_index = j
```

Comment:
nit: this can be computed faster, but probably not worth it.

Reply:
I agree that we should remove the logic involving `stop_strategy`, but we could use still use the random sampling strategy based on the following line:

```
self._weighted_samples = [int(w * len(d)) for w, d in zip(weights, datasets)]
self._normalized_probs = [ w / self._total_samples for w in self._weighted_samples ] 
iter_index = self._rng.choice(len(iters), p=self._normalized_probs)
```
This would ensure that on average the samples following the datasets would follow the distribution according to `self._normalized_probs`.

Reply:
But we'll get that with the new code even without the random sampling. It will follow the probabilities in deterministic fashioin.

Reply:
It is not clear to me what the benefit is to follow probabilities in deterministic fashion. If we want create random batches, we should allow random sampling of subsets, regardless of whether it would materially change the quality of the trained model. A deterministic order can be achieved by choosing the seed of the random generator.

I am also okay with using a deterministic order. Just thought that there is a cleaner alternative.

Reply:
But the randomness here is just the iterator selection, i.e., what dataset is sampled from next. This adds a small amount of noise but still overall selects the same samples. What's the real value in that?

Reply:
I agree that deterministic and random methods are the same. The difference maker is the change from "for loop" to "while True till exhausted".

I do think that the "while True" method had other issues (e.g. the distribution of data will change when one dataset is exhausted before others). I assume that ZQ had accepted that trade-off with the previous code.

Reply:
I was mainly thinking from the angle of why we need to change the 'random sampling' method to the 'deterministic' method. I think I may have answered the question in the other post.